core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/TxnManager.java

```diff
@@ -0,0 +1,253 @@
+import org.lwjgl.PointerBuffer;
+import org.lwjgl.system.MemoryStack;
+
+/**
+ * Manager for LMDB transactions.
+ */
+class TxnManager {
+
+	private final Mode mode;
+	private final Set<Txn> active = new HashSet<>();
+	private final long[] pool;
+	private final StampedLock lock = new StampedLock();
+	private long env;
+	private volatile int poolIndex = -1;
+
+	TxnManager(long env, Mode mode) {
+		this.env = env;
+		this.mode = mode;
+		this.pool = mode == Mode.RESET ? new long[128] : null;
+	}
+
+	private long startReadTxn() throws IOException {
+		long readTxn;
+		try (MemoryStack stack = stackPush()) {
+			PointerBuffer pp = stack.mallocPointer(1);
+			E(mdb_txn_begin(env, NULL, MDB_RDONLY, pp));
+			readTxn = pp.get(0);
+		}
+		return readTxn;
+	}
+
+	/**
+	 * Wraps an existing transaction into a txn reference object.
+	 *
+	 * @param txn the existing read or write transactions
+	 * @return the txn reference object
+	 */
+	Txn createTxn(long txn) {
+		return new Txn(txn) {
+			@Override
+			public void close() {
+				// do nothing
+			}
+		};
+	}
+
+	/**
+	 * Creates a new read-only transaction reference.
+	 *
+	 * @return the new transaction reference
+	 * @throws IOException if the transaction cannot be started for some reason
+	 */
+	Txn createReadTxn() throws IOException {
+		Txn txnRef = new Txn(createReadTxnInternal());
+		synchronized (active) {
+			active.add(txnRef);
+		}
+		return txnRef;
+	}
+
+	long createReadTxnInternal() throws IOException {
+		long txn = 0;
+		if (mode == Mode.RESET) {
+			synchronized (pool) {
+				if (poolIndex >= 0) {
+					txn = pool[poolIndex--];
+				}
+			}
+			if (txn == 0) {
+				txn = startReadTxn();
+			} else {
+				mdb_txn_renew(txn);
+			}
+		} else {
+			txn = startReadTxn();
+		}
+		return txn;
+	}
+
+	<T> T doWith(Transaction<T> transaction) throws IOException {
+		long stamp = lock.readLock();
+		T ret;
+		try (MemoryStack stack = stackPush()) {
+			try (Txn txn = createReadTxn()) {
+				ret = transaction.exec(stack, txn.get());
+			}
+		} finally {
+			lock.unlockRead(stamp);
+		}
+		return ret;
+	}
+
+	/**
+	 * This lock is used to globally block all transactions by using a {@link StampedLock#writeLock()}. This is required
+	 * to block transactions while automatic resizing the memory map.
+	 *
+	 * @return lock for managed transactions
+	 */
+	StampedLock lock() {
+		return lock;
+	}
+
+	void activate() throws IOException {
+		synchronized (active) {
+			for (Txn txn : active) {
+				txn.setActive(true);
+			}
+		}
+	}
+
+	void deactivate() throws IOException {
+		synchronized (active) {
+			for (Txn txn : active) {
+				txn.setActive(false);
+			}
+		}
+	}
+
+	void reset() throws IOException {
+		synchronized (active) {
+			for (Txn txn : active) {
+				txn.reset();
+			}
+		}
+	}
+
+	enum Mode {
+		RESET,
+		ABORT,
+		NONE
+	}
+
+	class Txn implements Closeable, AutoCloseable {
+
+		long txn;
+		List<Long> staleTxns;
+		private volatile long version;
```

Comment:
Doesn't need to be volatile since it is only written to from code holding a write lock and read from code holding a read/write lock.

Reply:
`version` is written in `setActive()` and read in `version()` that themselves are not synchronized. Doesn't this imply the use of volatile here?

Reply:
You only call those methods while inside a read or write-locked section. The StampedLock guarantees happens-before relationships.

Reply:
Yes, you are right. I just thought that there is also the possibility to call `version()` outside of a locked section.
Do you think it is faster if `volatile` is removed?

Reply:
Might be ever so slightly faster. One issue with using volatile here is that you at some point call `version++` which is a non-atomic operation on a volatile variable. So you are already assuming that there needs to be some kind of external locking to prevent race conditions.

Reply:
I've removed the volatile modifier.